source/common/stats/utility.h

```diff
@@ -206,5 +207,55 @@ class Utility {
                                                StatNameTagVectorOptConstRef tags = absl::nullopt);
 };
 
+/**
+ * Holds a reference to a stat by name. Note that the stat may not be created
+ * yet at the tome CachedReference is created. Calling find() then does a lazy
+ * lookup, potentially returning absl::nullopt if the stat doesn't exist yet.
+ * StatReference works whether the name was constructed symbolically, or with
+ * StatNameDynamicStorage.
+ *
+ * Lookups are very slow, taking time proportional to the size of the scope,
+ * holding mutexes during the lookup. However once the lookup succeeds, the
+ * result is cached atomically, and further calls to find() are thus fast and
+ * mutex-free. The implementation may be faster for stats that are named
+ * symbolically.
+ *
+ * CachedReference is valid for the lifetime of the Scope. When the Scope
+ * becomes invalid, CachedReferences must also be dropped as they will hold
+ * pointers into the scope.
+ */
+template <class StatType> class CachedReference {
+public:
+  CachedReference(Scope& scope, absl::string_view name) : scope_(scope), name_(std::string(name)) {}
+
+  /**
+   * Finds the named stat, if it exists, returning it as an optional.
+   */
+  absl::optional<std::reference_wrapper<StatType>> find() {
+    StatType* stat = stat_.get([this]() -> StatType* {
```

Comment:
Seems like find() should really only be called once. Can we assert somehow that `stat_` has never been set?

Reply:
The intended semantics here are that `find()` does caching itself so it's fine to call it lots of times.

I'm not clear on what you are suggesting as an alternative?

Reply:
I just meant that after you call find() once, the result is cached. Sorry I think I didn't give a full comment. Should there be some different between initialize() and find() in that initialize() is called once and find() should be called after initialize()? I don't feel strongly about it. In that case you could s/find/get.

Reply:
changed to `get()`